### DISCUSSION TITLE
Prometheus config file updates have noticeable large swing 33k -> 2.4mb

### DISCUSSION THREAD

**Discussion by vhadianto** [prometheus-operator/prometheus-operator#5297] [Q&A]
We have a few Prometheus instances, each has its own Operator.

Two or our Prometheus instances we have a few hundred Service Monitors (300+). We notice that (presumably) the Prometheus Operator keeps updating this file here:

```
/etc/prometheus/config_out/prometheus.env.yaml
```

This file currently sits at 2.4mb but occasionally this drops to 33k and then back to 2.4mb. When it drops to 33k the only targets in this file are Prometheus itself and none of our Service Monitors. I'm assuming that this is when there's a change in the Service Monitor. We have a rather dynamic environment where Service Monitors are created & updated semi frequently (it's a multi tenant system where tenants are created/destroyed dynamically by the users).

Is this expected? It seems like when the Operator is regenerating the config, there's a lag when thee new file is ready and during that time a "default" file is generated.

Anyone knows what's happening here or has some idea what might be going on? Or perhaps a better way to setup our Prometheus?

Thanks

**Comment by simonpasquier** [comment]
Your analysis is mostly correct: the prometheus operator creates/updates the configuration Secret (`prometheus-<prometheus resource name>`) and the config-reloader container will expand + copy the final configuration to `/etc/prometheus/config_out/prometheus.env.yaml`.
I'd check the logs of the prometheus operator and config-reloader container to see if there are any warning/error.

**Comment by omerosaienni** [comment]
The memory issue was coming from how Prometheus Operator was configured. 

For the first instance, we had the following configuration: 

Shard Prometheus installed in namespace: `prometheus-shard` (The one exhibiting the behaviour)

```sh
--namespaces=prometheus-shard,client-namespace
```

But for our main Prometheus instance at namespace `prometheus-main`, we had the following configuration:

```sh
--deny-namespaces=client-namespace
```

I figured out that the Operator Pod running in the namespace `prometheus-main` was interacting with the Prometheus Instance running at namespace `prometheus-shard`.

Adding the following to the configuration stop the problems we saw:

```sh
--deny-namespaces=client-namespace,prometheus-shard
```

There was nothing obvious that we saw in the logs.
